tests/sparse_test.py

```diff
@@ -131,7 +133,78 @@ def _is_required_cuda_version_satisfied(cuda_version):
     return int(version.split()[-1]) >= cuda_version
 
 
-class cuSparseTest(jtu.JaxTestCase):
+class SparseTestCase(jtu.JaxTestCase):
+  def assertSparseArraysEquivalent(self, x, y, *, check_dtypes=True, atol=None,
+                                   rtol=None, canonicalize_dtypes=True, err_msg=''):
+    x_bufs, x_tree = tree_util.tree_flatten(x)
+    y_bufs, y_tree = tree_util.tree_flatten(y)
+
+    self.assertEqual(x_tree, y_tree)
+    self.assertAllClose(x_bufs, y_bufs, check_dtypes=check_dtypes, atol=atol, rtol=rtol,
+                        canonicalize_dtypes=canonicalize_dtypes, err_msg=err_msg)
+
+  def _CheckAgainstDense(self, dense_op, sparse_op, args_maker,
+                         check_dtypes=True, tol=None, atol=None, rtol=None,
+                         canonicalize_dtypes=True):
+    """Check an operation against a dense equivalent"""
+    is_sparse = lambda x: isinstance(x, sparse.JAXSparse)
+
+    sparse_args = args_maker()
+    dense_args = tree_util.tree_map(sparse.todense, sparse_args, is_leaf=is_sparse)
+
+    sparse_ans = sparse_op(*sparse_args)
+    actual = tree_util.tree_map(sparse.todense, sparse_ans, is_leaf=is_sparse)
+    expected = dense_op(*dense_args)
+
+    self.assertAllClose(expected, actual, check_dtypes=check_dtypes,
+                        atol=atol or tol, rtol=rtol or tol,
+                        canonicalize_dtypes=canonicalize_dtypes)
+
+  def _CompileAndCheckSparse(self, fun, args_maker, check_dtypes=True,
+                             rtol=None, atol=None, check_cache_misses=True):
+    args = args_maker()
+
+    def wrapped_fun(*args):
+      # TODO(jakevdp): this check fails for sparse inputs. Why?
```

Comment:
Any guesses? Should we try and answer this as part of this PR?

Reply:
Ah, I figured it out: it's because `args_maker` is creating buffers whose size differs depending on the random seed, so it's triggering a cache miss the second time the args are created.

Reply:
I'll work on a fix...

Reply:
Done, please take a look!